pkg/serviceaccount/tokens_controller.go

```diff
@@ -240,8 +244,16 @@ func (e *TokensController) secretDeleted(obj interface{}) {
 		return
 	}
 
-	if _, err := e.removeSecretReferenceIfNeeded(serviceAccount, secret.Name); err != nil {
-		glog.Error(err)
+	for i := 1; i <= NumServiceAccountRemoveReferenceRetries; i++ {
```

Comment:
If the controller is stopped, and then a secret is deleted, and then the controller restarts, this code won't see the delete event.  

Therefore, to be really reliable, the controller needs to relist all ServiceAccounts upon startup and see if there are any that have dangling references to secrets, and removeSecretReferenceIfNeeded.

Given that it has to do this at startup, and that it need to be restartable, you might as well just run this check periodically.  Which obviates the need for the code below.